packages/loader/container-loader/src/container.ts

```diff
@@ -815,17 +817,16 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
                 this._attachState = AttachState.Attaching;
                 this.emit("attaching");
```

Comment:
I'd remove this event and replace it with a function call (this.context.notifyAttaching() - see we subscribe for this even in ContainerContext).
It's very easy to miss that this event is critical to the whole flow and that raising this event before capturing summary is going to break the flow (for case where this.loader.services.detachedBlobStorage.size > 0).

You would need to move that event / call into IF block above and replicate for new flow where we capture summary

Reply:
>It's very easy to miss that this event is critical to the whole flow and that raising this event before capturing summary is going to break the flow 

What breaks?

If we only want to capture the summary before "attaching" event, doesn't that mean we will have to support uploading new blobs through (mostly) the entire attach process, since creating the empty file and uploading blobs then happen in detached state?

Reply:
This event indicates to runtime when it switched to Attaching state and starts generating ops for local changes.
It is assumed (in today's workload) that switching to attaching state happens immediately after capturing summary (ops will be flushed later, when file is finally created).

Having any async gap between these two steps will result in bugs as data would be inconsistent.

Reply:
@vladsud But how does function is better than an event in terms of importance. I think event was fine because it is supposed to be called only once and in ContainerContext.ts too we were listening as event.once. ?

Reply:
I believe there is no need to have that event on public API surface. I.e. if the only consumer is ContainerContext, then Container should call an API on ContainerContext directly, same way ContainerContext calls into IRuntime.
That said, I'd even go father, and remove it altogether by adding attaching: boolean argument to createSummary() to indicate that runtime needs to switch to runtime, as conversion to attaching always happens at the same time as capturing summary (reverse is not true - we do capture in serialize() as well).